File: 2024/python/day1.py
import collections
from pathlib import Path


def main():
    with open(Path(__file__).resolve().parents[1] / "data" / "day1.txt") as infile:
        l_arr: list[int] = []
        r_arr: list[int] = []
        for line in infile.readlines():
            l, r = list(map(int, line.rstrip().split()))
            l_arr.append(l)
            r_arr.append(r)

        l_arr = sorted(l_arr)
        r_arr = sorted(r_arr)
        result = 0
        for l, r in zip(l_arr, r_arr):
            result += abs(l - r)
        print(result)

        counter = collections.Counter(r_arr)
        result2 = 0
        for l in l_arr:
            result2 += l * counter[l]
        print(result2)


if __name__ == "__main__":
    main()

File: 2024/rust/src/day1.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

pub fn main() {
    let file_path = std::env::current_dir()
        .unwrap()
        .parent()
        .unwrap()
        .join("data/day1.txt");

    if let Ok(lines) = read_lines(file_path) {
        let lines: Vec<(i32, i32)> = lines
            .flatten()
            .map(|line| {
                let (l, r) = line.split_once("   ").unwrap();
                (l.parse().unwrap(), r.parse().unwrap())
            })
            .collect();

        let (mut l_arr, mut r_arr): (Vec<i32>, Vec<i32>) = lines.into_iter().unzip();
        l_arr.sort();
        r_arr.sort();

        let result = l_arr
            .iter()
            .zip(&r_arr)
            .fold(0, |acc, (l, r)| acc + (l - r).abs());
        println!("{}", result);

        let result2 = l_arr.iter().fold(0, |acc, i| {
            let count = r_arr.iter().filter(|&n| n == i).count();
            acc + i * count as i32
        });
        println!("{}", result2);
    }
}

fn read_lines<P>(path: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(path)?;
    Ok(io::BufReader::new(file).lines())
}
